src/3D/Camera.cpp

```diff
@@ -172,39 +228,106 @@ void Camera::handleMouseInput(const SDL_Event& e)
 	// Holding down the middle mouse button enables free look.
 	if (e.type == SDL_MOUSEMOTION && e.motion.state & SDL_BUTTON(SDL_BUTTON_MIDDLE))
 	{
-		glm::vec3 rot = GetRotation();
-
-		rot.y -= e.motion.xrel * _freeLookSensitivity * 0.1f;
-		rot.x -= e.motion.yrel * _freeLookSensitivity * 0.1f;
-
-		SetRotation(rot);
+		int width = 1920;
+		int height = 1080;
+		Game::instance()->GetWindow()->GetSize(width, height);
+		float yaw = e.motion.xrel * (glm::two_pi<float>() / width);
+		float pitch = e.motion.yrel * (glm::pi<float>() / height);
+
+		glm::mat4x4 rotationMatrixX(1.0f);
+		rotationMatrixX = glm::rotate(rotationMatrixX, yaw, glm::vec3(0.0f, 1.0f, 0.0f));
+		_position = (rotationMatrixX * glm::vec4(_position - _viewCenterLand, 0.0f)) + glm::vec4(_viewCenterLand, 0.0f);
+
+		glm::mat4x4 rotationMatrixY(1.0f);
+		rotationMatrixY = glm::rotate(rotationMatrixY, pitch, GetRight());
+		_position = (rotationMatrixY * glm::vec4(_position - _viewCenterLand, 0.0f)) + glm::vec4(_viewCenterLand, 0.0f);
+
+		_rotation.x -= pitch;
+		_rotation.y -= yaw;
 	}
 	else if (e.type == SDL_MOUSEMOTION && e.motion.state & SDL_BUTTON(SDL_BUTTON_LEFT))
 	{
 		const auto& land = Game::instance()->GetLandIsland();
+
 		auto momentum = _position.y / 300;
-		auto forward = GetForward() * static_cast<float>(e.motion.yrel * momentum);
+		auto forward = GetForward() * glm::vec3(1.f, 0.f, 1.f) * static_cast<float>(e.motion.yrel * momentum);
 		auto right = GetRight() * -static_cast<float>(e.motion.xrel * momentum);
 		auto futurePosition = _position + forward + right;
-		auto height = land.GetHeightAt(glm::vec2(futurePosition.x + 5, futurePosition.z + 5)) + 13.0f;
+		auto height = land.GetHeightAt(glm::vec2(futurePosition.x + 5, futurePosition.z + 5)) + _groundHeightStart;
+		auto maxHandVel = std::max(log(height) * 10.0f, 5.0f);
 		futurePosition.y = std::max(height, _position.y);
-		_position = futurePosition;
+		glm::vec3 vecTo = futurePosition - _position;
+		vecTo.y = 0;
+		glm::mat3 mRotation = glm::transpose(GetRotationMatrix());
+		if (glm::length(vecTo) > maxHandVel)
+		{
+			vecTo = glm::normalize(vecTo) * maxHandVel;
+		}
+		_dv += vecTo * mRotation * 0.1f;
+	}
+	else if (e.type == SDL_MOUSEBUTTONUP && e.button.button == SDL_BUTTON(SDL_BUTTON_LEFT))
+	{
+		_dv = glm::vec3(0.0f);
+	}
+	else if (e.type == SDL_MOUSEBUTTONDOWN && e.button.button == SDL_BUTTON(SDL_BUTTON_LEFT))
+	{
+		_groundHeightStart = GetGroundPlaneAltitude();
+	}
+	else if (e.type == SDL_MOUSEBUTTONDOWN && e.button.button == SDL_BUTTON(SDL_BUTTON_MIDDLE))
+	{
+		// get the _viewCenterLand by raycasting to the land down the camera ray
+		float intersectDistance = 0.0f;
+		glm::vec3 viewVec = GetForward();
+		if (glm::intersectRayPlane(_position, viewVec, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f),
+		                           intersectDistance))
```

Comment:
This is very similar to `GetGroundPlaneAltitude`, maybe you can make it multipurpose.

Reply:
As I've now got rid of `GetGroundPlaneAltitude` I've left this in as a stand alone call to `glm::intersectRayPlane`